mavlink/message.go

```diff
@@ -102,53 +103,91 @@ func newPacketFromBytes(b []byte) (*Packet, int) {
 // a message they're interested in, and convert it to the
 // corresponding type via Message.FromPacket()
 func (dec *Decoder) Decode() (*Packet, error) {
-
-	// discard bytes until our start byte
 	for {
-		c, err := dec.br.ReadByte()
-		if err != nil {
-			return nil, err
+		startFoundInBuffer := false
+		// discard bytes in buffer before start byte
+		for i, b := range dec.buffer {
+			if b == startByte {
+				dec.buffer = dec.buffer[i:]
+				startFoundInBuffer = true
+				break
+			}
 		}
-		if c == startByte {
-			break
+
+		// if start not found, read until we see start byte
+		if !startFoundInBuffer {
+			for {
+				c, err := dec.br.ReadByte()
+				if err != nil {
+					return nil, err
+				}
+				if c == startByte {
+					dec.buffer = append(dec.buffer, c)
+					break
+				}
+			}
 		}
-	}
 
-	// hdr contains LENGTH, SEQ, SYSID, COMPID, MSGID
-	hdr := make([]byte, 5)
-	if _, err := io.ReadFull(dec.br, hdr); err != nil {
-		return nil, err
-	}
+		if len(dec.buffer) < 2 {
+			// read length byte
+			bytesRead := make([]byte, 1)
```

Comment:
Possible to read directly into the buffer without allocating another slice?  Not sure how much of a performance impact this makes, maybe it's not a big deal.